scripts/ai/PathfinderController.lua

```diff
@@ -0,0 +1,364 @@
+	self.vehiclesToIgnore = vehiclesToIgnore
+	self.mustBeAccurate = mustBeAccurate
+	self.areaToIgnoreFruit = areaToIgnoreFruit
+end
+
+--- Disables the fruit avoidance
+function PathfinderControllerContext:ignoreFruit()
+	self.maxFruitPercent = math.huge
+end
+
+function PathfinderControllerContext:getNumRetriesAllowed()
+	return self.numRetries
+end
+
+function PathfinderControllerContext:__tostring()
+	local str = "PathfinderControllerContext(vehicle name=%s, maxFruitPercent=%s, "+ 
+		"offFieldPenalty=%s, useFieldNum=%s, areaToAvoid=%s, allowReverse=%s, "+
+		"vehiclesToIgnore=%s, mustBeAccurate=%s, areaToIgnoreFruit=%s)"
+	return string.format(str,
+		CpUtil.getName(self.vehicle),
+		tostring(self.maxFruitPercent),
+		tostring(self.offFieldPenalty),
+		tostring(self.useFieldNum),
+		tostring(self.areaToAvoid),
+		tostring(self.allowReverse),
+		tostring(self.vehiclesToIgnore),
+		tostring(self.mustBeAccurate),
+		tostring(self.areaToIgnoreFruit))
+end
+
+---@class DefaultFieldPathfinderControllerContext : PathfinderControllerContext
+DefaultFieldPathfinderControllerContext = CpObject(PathfinderControllerContext)
+
+function DefaultFieldPathfinderControllerContext:init(...)
+	PathfinderControllerContext.init(self, ...)
+end
+
+---@class PathfinderController 
+PathfinderController= CpObject()
+
+PathfinderController.SUCCESS_FOUND_VALID_PATH = 0
+PathfinderController.ERROR_NO_PATH_FOUND = 1
+PathfinderController.ERROR_INVALID_GOAL_NODE = 2
+function PathfinderController:init(vehicle)
+	self.vehicle = vehicle
+	---@type PathfinderInterface
+	self.pathfinder = nil
+	---@type PathfinderControllerContext
+	self.lastContext = nil
+	self:reset()
+end
+
+function PathfinderController:__tostring()
+	return string.format("PathfinderController(failCount=%d, numRetries=%s, active=%s)",
+		self.failCount, self.numRetries, tostring(self.pathfinder ~= nil))
+end
+
+function PathfinderController:reset()
+	self.numRetries = 0
+	self.failCount = 0
+	self.startedAt = 0
+	self.timeTakenMs = 0
+	self.callbackClass = nil
+	self.callbackSuccessFunction = nil 
+	self.callbackRetryFunction = nil
+	self.lastContext = nil
+end
+
+function PathfinderController:update(dt)
+	if self:isActive() then
+		--- Applies coroutine for path finding
+		local done, path, goalNodeInvalid = self.pathfinder:resume()
+        if done then
+			self:finished(path, goalNodeInvalid)
+        end
+	end
+end
+
+function PathfinderController:isActive()
+	return self.pathfinder and self.pathfinder:isActive()
+end
+
+---@return PathfinderControllerContext
+function PathfinderController:getLastContext()
+	return self.lastContext
+end
+
+--- Sets the callbacks which get called on the pathfinder finish.
+---@param class table
+---@param successFunc function func(PathfinderController, success, path, goalNodeInvalid)
+---@param retryFunc function func(PathfinderController, last context, was last retry, retry attempt number)
+function PathfinderController:setCallbacks(class, successFunc, retryFunc)
```

Comment:
I like to use registerListeners() and notifyListeners() but this is really a matter of taste